### PR TITLE
Make mAP 3D more general

### PR DESCRIPTION
**Pull request by kprokofi**
### Summary

- [x] Remove AP2D from custom KITTI implementation and use torch metrics instead
- [x] Remove unnecessary hardcoded parts for KITTI benchmark and make it more general, add AP@IoU0.5 as target metric
- [x]  Fix some comments from Eunwoo: https://github.com/openvinotoolkit/training_extensions/pull/3979 

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by kprokofi** [comment]
Current limitations:
Core IoU implementation used by KITTI benchmark and by many opensource projects are not trustworthy. While it provides expected result on the trained snapshot, the IoU between exactly same boxes are 0. 

Next step in the future: try to substitute box_overlap core function with https://pytorch3d.org/docs/iou3d

**Review by harimkang — DISMISSED** [review]

File: src/otx/core/metrics/average_precision_3d.py [hunk 1]
```diff
@@ -32,6 +34,7 @@ def __init__(
         super().__init__()
 
         self.label_info: LabelInfo = label_info
+        self.mean_ap: MeanAveragePrecision = MeanAveragePrecision(box_format="xyxy", iou_type="bbox")
```
Comment (eunwoosh) [inline src/otx/core/metrics/average_precision_3d.py:37]: How about adding a code to reset `mean_ap` in `reset` function?

File: src/otx/core/metrics/kitti_3d_eval/eval.py [hunk 2]
```diff
@@ -65,12 +33,12 @@ def clean_data(
...
 
     Returns:
-        tuple: The number of valid objects, ignored_gt, ignored_dt, and dc_bboxes.
+        tuple: The number of valid objects, ignored_gt, ignored_dt.
     """
-    min_height = [40, 25, 25]
-    max_occlusion = [0, 1, 2]
-    max_truncation = [0.15, 0.3, 0.5]
-    dc_bboxes, ignored_gt, ignored_dt = [], [], []
+    min_height = 20
```
Comment (eunwoosh) [inline src/otx/core/metrics/kitti_3d_eval/eval.py:38]: According to value below, it should be set as 25. Is it intended to set it to 20?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work. I left some comments. Please take a look.

File: src/otx/core/metrics/kitti_3d_eval/eval.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/core/metrics/kitti_3d_eval/eval.py:38]: I decided to soft the limits of data filtering and keep more boxes. Previously, we had so called difficulties defined by KITTI benchmark, I removed them now and keep only one number to filter really bad GTs and predictions. I also removed truncation and occlusion, since this info is not available in other datasets and Geti.

File: src/otx/core/metrics/average_precision_3d.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/core/metrics/average_precision_3d.py:37]: Added https://github.com/openvinotoolkit/training_extensions/pull/4031/commits/779cc8f3c3e8682d264b56f2344ea2e5f3a5a70a

**Review by eunwoosh — DISMISSED** [review]

File: src/otx/algo/object_detection_3d/losses/ddn_loss.py [hunk 3]
```diff
@@ -76,12 +76,11 @@ def forward(
...
         Args:
-            loss [torch.Tensor(B, H, W)]: Pixel-wise loss
-            gt_boxes2d [torch.Tensor (B, N, 4)]: 2D box labels for foreground/background balancing
+            loss (torch.Tensor(B, H, W)): Pixel-wise loss
+            gt_boxes2d (torch.Tensor (B, N, 4)): 2D box labels for foreground/background balancing
 
         Returns:
-            loss [torch.Tensor(1)]: Total loss after foreground/background balancing
-            tb_dict [dict[float]]: All losses to log in tensorboard
+            loss (torch.Tensor(1)): Total loss after foreground/background balancing
```
Comment (sungchul1) [inline src/otx/algo/object_detection_3d/losses/ddn_loss.py]: Is it normal to insert shape in type hint?

File: src/otx/algo/object_detection_3d/losses/monodetr_loss.py [hunk 4]
```diff
@@ -29,11 +29,10 @@ def __init__(self, num_classes: int, weight_dict: dict, focal_alpha: float, grou
...
         Args:
-            num_classes: number of object categories, omitting the special no-object category
-            matcher: module able to compute a matching between targets and proposals
-            weight_dict: dict containing as key the names of the losses and as values their relative weight.
-            focal_alpha: alpha in Focal Loss
-            group_num: number of groups for data parallelism
+            num_classes (int): number of object categories, omitting the special no-object category
+            weight_dict (dict): dict containing as key the names of the losses and as values their relative weight.
+            focal_alpha (float): alpha in Focal Loss
+            group_num (int): number of groups for data parallelism
```
Comment (sungchul1) [inline src/otx/algo/object_detection_3d/losses/monodetr_loss.py]: Some have ".", some don't. Please unify.

File: src/otx/core/metrics/average_precision_3d.py [hunk 5]
```diff
@@ -51,13 +55,35 @@ def update(self, preds: list[dict[str, Tensor]], target: list[dict[str, Tensor]]
...
+        targets_for_torchmetrics = self.prepare_inputs_for_map_coco(self.targets)
+        ap_bbox_coco = self.mean_ap(preds_for_torchmetrics, targets_for_torchmetrics)
+        ap_3d = get_coco_eval_result(
             self.targets,
             self.preds,
             current_classes=[curcls.lower() for curcls in current_classes],
         )
-        # use moderate difficulty as final score. Average across all calsses.
-        return {"mAP_bbox_3d": Tensor([map_3d[:, 1].mean()]), "mAP_bbox_2d": Tensor([map_bbox[:, 1].mean()])}
+        # Average across all calsses.
```
Comment (sungchul1) [inline src/otx/core/metrics/average_precision_3d.py]:
```suggestion
        # Average across all classes.
```

File: src/otx/core/metrics/kitti_3d_eval/eval.py
Comment (sungchul1) [inline src/otx/core/metrics/kitti_3d_eval/eval.py:1]: This file seems to need test.

**Review by sungchul1 — COMMENTED** [review]
LGTM, but I left some minor comments.
If they can be updated in the next PR, I'll approve it.

File: src/otx/algo/object_detection_3d/losses/ddn_loss.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/algo/object_detection_3d/losses/ddn_loss.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/4031/commits/8c2576f094dc7630e6d26333cddebb785d42d76f

File: src/otx/algo/object_detection_3d/losses/monodetr_loss.py (hunk 4 above)
Comment (kprokofi) [reply src/otx/algo/object_detection_3d/losses/monodetr_loss.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/4031/commits/8c2576f094dc7630e6d26333cddebb785d42d76f

File: src/otx/core/metrics/kitti_3d_eval/eval.py
Comment (kprokofi) [reply src/otx/core/metrics/kitti_3d_eval/eval.py:1]: Actually, I intentionally don't add any tests right now, because 3d IoU calculation cannot pass meaningful test. We already have integration test to check if metric can be computed. So, I don't see the point to add tests just to check if it can be launched, I will get back to this after reviewing 3d IoU once again.

File: src/otx/core/metrics/average_precision_3d.py (hunk 5 above)
Comment (kprokofi) [reply src/otx/core/metrics/average_precision_3d.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/4031/commits/8c2576f094dc7630e6d26333cddebb785d42d76f

**Review by eunwoosh — APPROVED** [review]

**Review by sovrasov — APPROVED** [review]
